### PR TITLE
Subfolder sharing info

### PR DESCRIPTION
**Pull request by jvillafanez**
## Description
Return the share information inside the shared folders.

There is a new property `<oc:share-types-parents />` to request this information in the propfind
```
<?xml version="1.0"?>
<d:propfind  xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
  <d:prop>
    <d:getlastmodified />
    <d:getetag />
    <d:getcontenttype />
    <d:resourcetype />
    <oc:fileid />
    <oc:permissions />
    <oc:size />
    <d:getcontentlength />
    <oc:tags />
    <oc:favorite />
    <d:lockdiscovery />
    <oc:comments-unread />
    <oc:owner-display-name />
    <oc:share-types />
    <oc:share-types-parents />
  </d:prop>
</d:propfind>
```
Example output:
```
      .......
      <d:prop>
        ....
        <d:lockdiscovery/>
        <oc:comments-unread>0</oc:comments-unread>
        <oc:owner-display-name>user1</oc:owner-display-name>
        <oc:share-types/>
        <oc:share-types-parents>
          <oc:parents>
            <oc:parent>
              <d:href>/remote.php/dav/files/user1/foo</d:href>
              <oc:share-type>0</oc:share-type>
            </oc:parent>
            <oc:parent>
              <d:href>/remote.php/dav/files/user1/foo/bar</d:href>
              <oc:share-type>3</oc:share-type>
            </oc:parent>
          </oc:parents>
        </oc:share-types-parents>
      </d:prop>
```

Checked with a very inner folder with 10000 files, the request takes around 13 secs average, against around 12 secs without the new property

```
curl -w "@curl-format.txt" -o /dev/null --data-binary @propfindtest1.xml -X PROPFIND -u user1:11111 'http://10.0.2.8:7080/remote.php/dav/files/user1/foo/bar/a/b/c/d/e/f/g/h/i/j/k/l/m/n/o/p/q/r/s/t/u/v/w/x/y/z/folder0'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 9951k    0 9950k  100   469   773k     36  0:00:13  0:00:12  0:00:01 2730k
    time_namelookup:  0,000031
       time_connect:  0,000605
    time_appconnect:  0,000000
   time_pretransfer:  0,000695
      time_redirect:  0,000000
 time_starttransfer:  12,748179
                    ----------
         time_total:  12,868428
```

It also include an optimization for the FS blacklist function because it was taking quite a bit of time and there are a lot of calls to that function.

## Related Issue

https://github.com/owncloud/enterprise/issues/3453

## How Has This Been Tested?

Manually checked with curl

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

File: lib/private/Files/Filesystem.php [hunk 1]
```diff
@@ -637,38 +637,36 @@ public static function isIgnoredDir($dir) {
...
-			$match = \array_intersect($path_parts, $excluded);
-			if ($match) {
-				return true;
+
+		$lowercaseFileOrDir = \strtolower($FileOrDir);
+		// explode the file or dir
+		foreach (\explode('/', $FileOrDir) as $part) {
+			if ($part !== '') {
+				// further explode each array element with '\' and add to result array if found
+				foreach (\explode('\\', $part) as $microPart) {
```
Comment (jvillafanez) [inline lib/private/Files/Filesystem.php]: can we remove this piece? I'm not sure if this is just for windows' paths.

**Comment by phil-davis** [comment]
@micbar this is not tagged `deferred`. Is it moving towards a release?

**Comment by micbar** [comment]
Yes, this belongs to https://github.com/owncloud/enterprise/issues/3453

**Comment by phil-davis** [comment]
@micbar there is no QA-team label here.
I guess we should add a couple of acceptance test cases that do PROPFIND with the new `oc:share-types-parents` and make sure something sensible is returned?

**Comment by individual-it** [comment]
assigned my-self to write acceptance tests

**Comment by phil-davis** [comment]
Not: CI here is 3 months old. 
@individual-it when you push some tests, it would be good to rebase to make sure...

**Comment by individual-it** [comment]
- [ ] href is not URL encoded

see this response (the folder name of the shared folder is `simple % folder`)
```
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/owncloud-core/remote.php/dav/files/admin/simple%20%25%20folder/data.zip</d:href>
    <d:propstat>
      <d:prop>
        <oc:share-types-parents>
          <oc:parents>
            <oc:parent>
              <d:href>/owncloud-core/remote.php/dav/files/admin/simple % folder</d:href>
              <oc:share-type>3</oc:share-type>
              <oc:share-type>0</oc:share-type>
              <oc:share-type>1</oc:share-type>
            </oc:parent>
          </oc:parents>
        </oc:share-types-parents>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

**Comment by individual-it** [comment]
- [ ] reshares are not reported to the original sharer
- [ ] original shares are not reported to a resharer

**Comment by individual-it** [comment]
@phil-davis please review acceptance tests
@jvillafanez @PVince81 please see if the acceptance tests meet the expectations, also there are issues/questions about expectations, see https://github.com/owncloud/core/pull/36123#issuecomment-563898458 & https://github.com/owncloud/core/pull/36123#issuecomment-563930404

**Comment by jvillafanez** [comment]
I think we decided to change the approach and make a fontend-only solution so I don't think this is relevant any longer.

> href is not URL encoded
> see this response (the folder name of the shared folder is simple % folder)

Yes, that's a bug.

File: tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    And user "user0" has created a share with settings
+      | path        | PARENT |
+      | shareType   | user   |
+      | permissions | all    |
+      | shareWith   | user1  |
+    When user "user0" gets the following properties of folder "/PARENT/CHILD" using the WebDAV API
+      | oc:share-types-parents |
+    Then the value of the item "//oc:share-types-parents/oc:parents/oc:parent/d:href" in the response should be "<href>"
+    And the value of the item "//oc:share-types-parents/oc:parents/oc:parent/oc:share-type" in the response should be "0"
+    When user "user0" gets the following properties of folder "/PARENT/CHILD/child.txt" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature]:
```suggestion
    When user "user0" gets the following properties of file "/PARENT/CHILD/child.txt" using the WebDAV API
```

File: tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    And user "user0" has created a share with settings
+      | path        | PARENT |
+      | shareType   | group  |
+      | permissions | all    |
+      | shareWith   | grp1   |
+    When user "user0" gets the following properties of folder "/PARENT/CHILD" using the WebDAV API
+      | oc:share-types-parents |
+    Then the value of the item "//oc:share-types-parents/oc:parents/oc:parent/d:href" in the response should be "<href>"
+    And the value of the item "//oc:share-types-parents/oc:parents/oc:parent/oc:share-type" in the response should be "1"
+    When user "user0" gets the following properties of folder "/PARENT/CHILD/child.txt" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature]:
```suggestion
    When user "user0" gets the following properties of file "/PARENT/CHILD/child.txt" using the WebDAV API
```

File: tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+      | new         | /%dav_path%/files/user0/PARENT |
+
+  Scenario Outline: Retrieving share-types-parents of items inside a folder that is shared by a public-link
+    Given using <dav_version> DAV path
+    And user "user0" has created a public link share of folder "/PARENT"
+    When user "user0" gets the following properties of folder "/PARENT/CHILD" using the WebDAV API
+      | oc:share-types-parents |
+    Then the value of the item "//oc:share-types-parents/oc:parents/oc:parent/d:href" in the response should be "<href>"
+    And the value of the item "//oc:share-types-parents/oc:parents/oc:parent/oc:share-type" in the response should be "3"
+    When user "user0" gets the following properties of folder "/PARENT/CHILD/child.txt" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature]:
```suggestion
    When user "user0" gets the following properties of file "/PARENT/CHILD/child.txt" using the WebDAV API
```

File: tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    Given using <dav_version> DAV path
+    And using server "REMOTE"
+    And user "user2" has been created with default attributes and skeleton files
+    And user "user0" from server "LOCAL" has shared "/PARENT" with user "user2" from server "REMOTE"
+    And using server "LOCAL"
+    When user "user0" gets the following properties of folder "/PARENT/CHILD" using the WebDAV API
+      | oc:share-types-parents |
+    Then the value of the item "//oc:share-types-parents/oc:parents/oc:parent/d:href" in the response should be "<href>"
+    And the value of the item "//oc:share-types-parents/oc:parents/oc:parent/oc:share-type" in the response should be "6"
+    When user "user0" gets the following properties of folder "/PARENT/CHILD/child.txt" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature]:
```suggestion
    When user "user0" gets the following properties of file "/PARENT/CHILD/child.txt" using the WebDAV API
```

File: tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+      | path        | PARENT |
+      | shareType   | group  |
+      | permissions | all    |
+      | shareWith   | grp1   |
+    And user "user0" has created a share with settings
+      | path        | PARENT |
+      | shareType   | user   |
+      | permissions | all    |
+      | shareWith   | user1  |
+    When user "user0" gets the following properties of file "/PARENT" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature]:
```suggestion
    When user "user0" gets the following properties of folder "/PARENT" using the WebDAV API
```

File: tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+      | oc:share-types-parents |
+    Then the response should contain an empty property "oc:share-types-parents/oc:parents"
+    Examples:
+      | dav_version |
+      | old         |
+      | new         |
+
+  Scenario Outline: Retrieving share-types-parents of a not shared item
+    Given using <dav_version> DAV path
+    When user "user0" gets the following properties of file "/PARENT" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature]:
```suggestion
    When user "user0" gets the following properties of folder "/PARENT" using the WebDAV API
```

File: tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+      | oc:share-types-parents |
+    Then the response should contain an empty property "oc:share-types-parents/oc:parents"
+    Examples:
+      | dav_version |
+      | old         |
+      | new         |
+
+  Scenario Outline: Retrieving share-types-parents of items inside a not shared folder
+    Given using <dav_version> DAV path
+    When user "user0" gets the following properties of file "/PARENT/CHILD" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature]:
```suggestion
    When user "user0" gets the following properties of folder "/PARENT/CHILD" using the WebDAV API
```

File: tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    And user "user1" has created a share with settings
+      | path        | PARENT (2) |
+      | shareType   | group      |
+      | permissions | all        |
+      | shareWith   | grp1       |
+    When user "user0" gets the following properties of folder "/PARENT/CHILD" using the WebDAV API
+      | oc:share-types-parents |
+    Then the value of the item "//oc:share-types-parents/oc:parents/oc:parent/d:href" in the response should be "<href>"
+    And the value of the item "//oc:share-types-parents/oc:parents/oc:parent/oc:share-type" in the response should be "0"
+    When user "user0" gets the following properties of folder "/PARENT/CHILD/child.txt" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature]:
```suggestion
    When user "user0" gets the following properties of file "/PARENT/CHILD/child.txt" using the WebDAV API
```

File: tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    And user "user1" has created a share with settings
+      | path        | PARENT (2) |
+      | shareType   | group      |
+      | permissions | all        |
+      | shareWith   | grp1       |
+    When user "user1" gets the following properties of folder "/PARENT (2)/CHILD" using the WebDAV API
+      | oc:share-types-parents |
+    Then the value of the item "//oc:share-types-parents/oc:parents/oc:parent/d:href" in the response should be "<href>"
+    And the value of the item "//oc:share-types-parents/oc:parents/oc:parent/oc:share-type" in the response should be "1"
+    When user "user1" gets the following properties of folder "/PARENT (2)/CHILD/child.txt" using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavProperties/subfolderSharingInfo.feature]:
```suggestion
    When user "user1" gets the following properties of file "/PARENT (2)/CHILD/child.txt" using the WebDAV API
```

**Review by phil-davis — COMMENTED** [review]
Tests look good. The (file|folder) thing in step text is annoying to keep consistent.
That is one reason to call it `resource` in the future.

**Comment by individual-it** [comment]
@micbar  @jvillafanez @pmaier1 closing this PR or investing more time into it?

**Comment by micbar** [comment]
@individual-it This is implemented on the Frontend, no need for this anymore.

**Pull request closed without merging**